aggregator/ref_count_buffer.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"sync/atomic"
+
+	"github.com/m3db/m3metrics/protocol/msgpack"
+)
+
+// RefCountedBuffer is a refcounted buffer.
+type RefCountedBuffer struct {
+	n   int32
+	buf msgpack.Buffer
+}
+
+// NewRefCountedBuffer creates a new refcounted buffer.
+func NewRefCountedBuffer(buffer msgpack.Buffer) *RefCountedBuffer {
+	return &RefCountedBuffer{n: 1, buf: buffer}
+}
+
+// Buffer returns the internal msgpack buffer.
+func (b *RefCountedBuffer) Buffer() msgpack.Buffer { return b.buf }
+
+// IncRef increments the refcount for the buffer.
+func (b *RefCountedBuffer) IncRef() int {
+	if n := int(atomic.AddInt32(&b.n, 1)); n > 0 {
+		return n
+	}
+	panic("invalid ref count")
+}
+
+// DecRef decrements the refcount for the buffer.
+func (b *RefCountedBuffer) DecRef() int {
```

Comment:
Seems like an implementation detail though, caller shouldn't care about whether we use int32 or uint32 or int64 internally to track the ref. The return type should be the general type used to count things (number of references in this case), which IMO should be int because that's what we use for `len` too.